bin/mchoose

```diff
@@ -0,0 +1,711 @@
+    var p, p_opts, p_choices, p_err, p_stats;
+
+    var cfgdata, cfg;
+    try {
+        cfgdata = mod_fs.readFileSync(opts.config, 'utf8');
+        cfg = JSON.parse(cfgdata);
+    } catch (e) {
+        cb(new VError('Unable to load %s: %s', opts.config, e.message));
+        return;
+    }
+
+    // Basic validation of the muskie/buckets-api config file.
+    if (!cfg.hasOwnProperty('storage')) {
+        cb(new VError('"storage" configuration not found'));
+        return;
+    }
+
+    function printChooseSummary(chooseStats) {
+        mod_assert.object(chooseStats, 'chooseStats');
+
+        var p_db = chooseStats.db;
+        var p_allDcs = Object.keys(p_db);
+        var p_dcsInUse = chooseStats.dcsInUse;
+        var p_offsets = chooseStats.offsets;
+
+        var summary = [];
+
+        summary.push('replicas requested: ' + replicas);
+        summary.push('size requested: ' + sizeMB + ' MB');
+
+        var allSharkCount = 0;
+        var inUseSharkCount = 0;
+        var sharksByDc = {};
+        p_allDcs.forEach(function (dc) {
+            var len = p_db[dc].length;
+            allSharkCount += len;
+            sharksByDc[dc] = {
+                count: len,
+                inUse: 0
+            };
+
+            var i = p_dcsInUse.indexOf(dc);
+            if (i !== -1) {
+                var c = p_db[dc].length - p_offsets[i];
+                sharksByDc[dc].inUse = c;
+                inUseSharkCount += c;
+            }
+        });
+
+        summary.push('storage server count by datacenter:');
+        p_allDcs.forEach(function (dc) {
+            summary.push('\t' + dc + ': ' + sharksByDc[dc].count +
+                ' (' + sharksByDc[dc].inUse + ' available)');
+        });
+        summary.push('total storage server count: ' + allSharkCount +
+            ' (' + inUseSharkCount + ' available)');
+
+        summary.push('');
+
+        if (p_err) {
+            summary.push('could not choose storage node set: ' +
+                p_err.cause().message);
+        } else {
+            mod_assert.object(p_choices, 'p_choices');
+            summary.push('choices:');
+
+            p_choices.forEach(function (c, i) {
+                summary.push(sprintf('\t%s:', i));
+                c.forEach(function (s, j) {
+                    summary.push(sprintf('\t\t%s (%s)', s['manta_storage_id'],
+                        s['datacenter']));
+                });
+
+                summary.push('');
+            });
+        }
+
+        console.log(summary.join('\n'));
+    }
+
+    mod_vasync.pipeline({
+        funcs: [
+        function readRecords(_, vcb) {
+            if (opts.file) {
+                mod_fs.readFile(opts.file, function (err, contents) {
+                    if (err) {
+                        err = new VError(err, 'reading %s', opts.file);
+                        vcb(err);
+                        return;
+                    }
+
+                    user_db = JSON.parse(contents);
```

Comment:
is it ok for this to error? or should this be caught and sent up the pipeline via the callback?

Reply:
Yeah, I added code to catch the exception and bail out with a useful error message.